src/js/legend.js

```diff
@@ -0,0 +1,241 @@
+        var table = that.table,
+            legendOptions = that.options.legendOptions,
+            colors = gpii.chartAuthoring.pieChart.legend.getColorArray(legendOptions.colors),
+            sort = legendOptions.sort;
+
+        // Consolidate user-supplied colors with dataset at draw time
+
+        var dataSet = gpii.chartAuthoring.pieChart.legend.addValueFromArray(that.model.dataSet, colors, "color");
+
+        var tbody = table.selectAll("tbody");
+
+        that.rows = tbody.selectAll("tr")
+                            .data(dataSet, function (d) {
+                                return d.id;
+                            });
+
+        gpii.chartAuthoring.pieChart.legend.addRows(that);
+
+        gpii.chartAuthoring.pieChart.legend.updateRows(that);
+
+        gpii.chartAuthoring.pieChart.legend.removeRows(that);
+
+        if (sort) {
+            that.rows.sort(that.sort);
+        }
+    };
+
+    gpii.chartAuthoring.pieChart.legend.create = function (that) {
+        var container = that.container,
+            tableClass = that.classes.table,
+            showLegendHeadings = that.options.legendOptions.showLegendHeadings;
+
+        that.table = that.jQueryToD3(container)
+            .append("table")
+            .attr({
+                "class": tableClass
+            });
+
+        that.table.append("thead");
+        that.table.append("tbody");
+        if(showLegendHeadings) {
+            var thead = that.table.selectAll("thead");
+
+            thead.append("th")
+                .attr({
+                    "scope":"col"
+                })
+                .html(that.options.strings.legendColHeading);
+
+            thead.append("th")
+                .attr({
+                    "scope":"col"
+                })
+                .html(that.options.strings.labelColHeading);
+
+            thead.append("th")
+                .attr({
+                    "scope":"col"
+                })
+                .html(that.options.strings.valueColHeading);
+        }
+
+        that.draw();
+
+        that.events.onLegendCreated.fire();
+    };
+
+    gpii.chartAuthoring.pieChart.legend.sort = function (a, b) {
+        return b.value - a.value;
+    };
+
+    gpii.chartAuthoring.pieChart.legend.getColorCellStyle = function (data) {
+        return "background-color: " + data.color + ";";
+    };
+
+    // Given an array of objects, an array of values and new value name, loop
+    // the object array in index order, apply the value from the array at the
+    // same index to the value name, and return a new object array with the
+    // added values
+
+    gpii.chartAuthoring.pieChart.legend.addValueFromArray = function (objectArray, valueArray, newValueName) {
+        return fluid.transform(objectArray, function (object, idx) {
+            var consolidated = fluid.copy(object);
+            // Don't do anything if not passed an actual array in the value array
+            if(fluid.isArrayable(valueArray)) {
```

Comment:
This check can be the very first thing of the function instead of inside a transform loop that gets called in every iteration.